protocol/storage/miner_test.go

```diff
@@ -297,18 +304,319 @@ func TestOnCommitmentAddedToChain(t *testing.T) {
+			t.Error("Should not have called submit post")
+			return [][]byte{}, nil
+		}
+		api.messageHandlers = handlers
+
+		height := uint64(50500)
+		api.blockHeight = types.NewBlockHeight(height)
+		block := &types.Block{Height: types.Uint64(height)}
+		ts, err := types.NewTipSet(block)
+		require.NoError(t, err)
+
+		miner.OnNewHeaviestTipSet(ts)
+
+		// OnNewHeaviestTipSet spawns a goroutine that should exit immediately dependencies are faked out
+		time.Sleep(1 * time.Second)
+
+		assert.Equal(t, 1, len(logger.ErrorMessages))
+		assert.Contains(t, logger.ErrorMessages[0], "too late start")
+	})
+}
+
+func successMessageHandlers(t *testing.T) messageHandlerMap {
+	handlers := messageHandlerMap{}
+	handlers["isBootstrapMiner"] = func(a address.Address, v types.AttoFIL, p ...interface{}) ([][]byte, error) {
+		return mustEncodeResults(t, false), nil
+	}
+
+	handlers["getSectorCommitments"] = func(a address.Address, v types.AttoFIL, p ...interface{}) ([][]byte, error) {
+		commitments := map[string]types.Commitments{}
+		commitments["42"] = types.Commitments{}
+		return mustEncodeResults(t, commitments), nil
+	}
+	handlers["getProvingPeriod"] = func(a address.Address, v types.AttoFIL, p ...interface{}) ([][]byte, error) {
+		return mustEncodeResults(t, types.NewBlockHeight(20003), types.NewBlockHeight(40003)), nil
+	}
+	handlers["submitPoSt"] = func(a address.Address, v types.AttoFIL, p ...interface{}) ([][]byte, error) {
+		return [][]byte{}, nil
+	}
+	return handlers
+}
+
+type testNode struct{}
+
+func (tn *testNode) BlockService() bserv.BlockService           { return nil }
+func (tn *testNode) Host() host.Host                            { return nil }
+func (tn *testNode) SectorBuilder() sectorbuilder.SectorBuilder { return &testSectorBuilder{} }
+
+type testSectorBuilder struct{}
+
+func (tsb *testSectorBuilder) AddPiece(ctx context.Context, pieceRef cid.Cid, pieceSize uint64, pieceReader io.Reader) (sectorID uint64, err error) {
+	return 0, nil
+}
+func (tsb *testSectorBuilder) ReadPieceFromSealedSector(pieceCid cid.Cid) (io.Reader, error) {
+	return nil, nil
+}
+func (tsb *testSectorBuilder) SealAllStagedSectors(ctx context.Context) error {
+	return nil
+}
+func (tsb *testSectorBuilder) SectorSealResults() <-chan sectorbuilder.SectorSealResult {
+	return nil
+}
+func (tsb *testSectorBuilder) GeneratePoSt(gpr sectorbuilder.GeneratePoStRequest) (sectorbuilder.GeneratePoStResponse, error) {
+	return sectorbuilder.GeneratePoStResponse{
+		Proofs: []types.PoStProof{[]byte("test proof")},
+	}, nil
+}
+func (tsb *testSectorBuilder) Close() error {
+	return nil
+}
+
+func mustEncodeResults(t *testing.T, results ...interface{}) [][]byte {
+	out := make([][]byte, len(results))
+	values, err := abi.ToValues(results)
+	require.NoError(t, err)
+
+	for i, value := range values {
+		encoded, err := value.Serialize()
+		require.NoError(t, err)
+		out[i] = encoded
+	}
+	return out
+}
+
+type testLogger struct {
```

Comment:
OBSERVATION: I think that all of this test code goes away if you change `OnNewHeaviestTipSet` to return an `error`.

Reply:
It did.